### PR TITLE
Jerry Hung Assignment 2

### PR DESCRIPTION
**Pull request by Jerry-Hung**

### CODE REVIEW

**Review by harrylyf — COMMENTED** [review]
- Did your partner find any interesting features of Scala that you didn't know about?
The importing does take a lot of time
- Did you find any interesting features of Scala that might help your partner?
We can define method inside method.
- Do your partners tests cover all the cases for sleepIn?
I did not find the test. But the code is right.
- Do you think your partner's changes to the Picture library make it easier for a user to compose operations on pictures?
Yes. I also use (.getWidth) to replace the unnecessary variables. And to make the return value (BufferedImage) is much easier. So we do not have to create the BufferedImage every time.
- Critique your partner's response in reflection.md. Do you have anything to add, and/or did their response help you see new or different things about the design of the Picture library?
Yes, I also find the code sometimes hard to follow. And there are many repeated operation that should be eliminated. I suggest we can talk about some tricks or shortcut about Scala in our next session.

**Comment by bwiedermann** [comment]
It looks like the tests for `sleepIn` in `PracticeTests.scala` are missing.

File: Practice.scala [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+  /**
+    * from http://codingbat.com/prob/p187868
+    *
+    * Returns true if we sleep in. We sleep in if it's not a weekday or
+    * we're on vacation.
+    */
+  def sleepIn(w: Boolean, v: Boolean): Boolean = w == false || v == true
+  // replace this comment with your implementation of sleepIn
+}
```
Comment (bwiedermann) [inline Practice.scala:21]: It looks like this file was in the top directory of the repository. It's redundant here (there's already a version of this file in `/src/main/scala/pioneer/practice`.

File: PracticeProgram.scala [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+object PracticeProgram extends App {
+  print("Enter a number: ")
+  val inputNumber = scala.io.StdIn.readInt()
+  val result = square(inputNumber)
+  println(s"The square of ${inputNumber} is ${result}.")
+  print("Enter a String: ")
+  val inputString = scala.io.StdIn.readLine()
+  val string = repeat(inputString)
+  println("repeating: " + string)
+}
```
Comment (bwiedermann) [inline PracticeProgram.scala:15]: It looks like this file was in the top directory of the repository. It's redundant here (there's already a version of this file in `/src/main/scala/pioneer/practice`.

File: reflection.md [hunk 3]
```diff
@@ -6,4 +6,4 @@ on the following question:
 Is there anything that is still annoying or difficult about the operations we
 have in the library (i.e., flipping, rotating, and grayscaling)?
 
-_replace this text with your answer_
+I still felt like the code itself is strenuous to understand, and I still felt like a lot of the program can be simplified in the sense that a lot of the code are "repeated".
```
Comment (bwiedermann) [inline reflection.md:9]: Yes, a lot of the `Picture` library implementation is repeated, and could be simplified. This question was also intended to get at the _user's_ perspective: From the user's point of view (e.g,. when writing `Practice.scala`, is there anything still annoying about the library?

Comment (bwiedermann) [reply reflection.md:9]: Additionally, this reflection could be improved with a bit more specificity: What in particular is strenuous to understand, and what parts can be simplified?

File: Picture.scala [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+/**
+  * A small, image-processing library. Thanks to
+  *    http://otfried.org/scala/image.html
+  * for help on how to perform basic image processing in Scala
+  */
+object Picture {
+
+  /** Flips an image along its horizontal axis */
+  def flipHorizontal(img: BufferedImage): BufferedImage = {
+    val image = img
```
Comment (bwiedermann) [inline Picture.scala:17]: This line works, but it isn't strictly necessary: You could name the parameter `image` and remove this line.

File: discussion-questions.md [hunk 5]
```diff
@@ -1,8 +1,9 @@
...
-_replace this text with your answer_
+Around 5 hours
 
 # Discussion questions / topics
 Below, add one or two things that you might want to discuss when we meet again.
 
-_replace this text with your discussion questions / topics_
\ No newline at end of file
+I think that Scala is very similar to java and python in the way the program runs.
+Also, the importing took a lot of time.
```
Comment (bwiedermann) [inline discussion-questions.md:9]: What you've written here are true observations, but could benefit from some specificity. In what ways are Scala, Java, and Python similar? By "importing", do you mean how to get Scala up and running (e.g., getting the project into IntelliJ)?

Also, the purpose of the discussion questions / topics is to seed conversation in our group meetings. From your description, it's difficult for me to know what you'd like to focus on in a group session. Would you have wanted to talk about how to import? Would you have wanted to talk about the similarities between Java, Python, and Scala?
